_posts/2024-04-01-oidc-proxy.adoc

```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: 'Use OIDC Proxy to integrate OIDC service endpoints with custom ChatGPT'
+date: 2024-04-01
+tags: extension oidc security chatgpt development-tips
+synopsis: 'Explain how OIDC Proxy can help to integrate OIDC service endpoints with custom ChatGPT'
+author: sberyozkin
+---
+:imagesdir: /assets/images/posts/oidc-proxy
+
+
+== Introduction
+
+https://github.com/quarkiverse/quarkus-oidc-proxy[Quarkus OIDC Proxy] is a new https://github.com/quarkiverse[Quarkiverse] extension which can help to integrate https://quarkus.io/guides/security-oidc-bearer-token-authentication[OIDC service endpoints] with external Single-page applications (SPA). SPA uses the OIDC authorization code flow itself (without relying on Quarkus) to authenticate the current user, and accesses the Quarkus OIDC service endpoint with the access token on behalf of the authenticated user.
+
+This is a simple diagram which shows how this process works, copied to this post from the https://quarkus.io/guides/security-oidc-bearer-token-authentication[OIDC Bearer token guide] for your convenience: 
+
+image::security-bearer-token-spa.png
+
+You can see that OIDC provider is used to authenticate the current user to SPA. SPA acquires ID and access tokens as part of the authorization code flow and uses the access token to access the Quarkus OIDC service endpoint. 
+
+SPA must know the OIDC provider connection details, including the registered OIDC application's client id, secret and other OIDC specific details required to complete the authorization code flow successfully. You must also allow an SPA specific callback URL in your registered OIDC application which may not always be acceptable.  
+
+https://github.com/quarkiverse/quarkus-oidc-proxy[Quarkus OIDC Proxy] emulates the OIDC provider in the diagram above. It interposes over Quarkus OIDC service endpoints and delegates to the real OIDC provider. It can help to integrate OIDC service endpoints with SPA without having to expose the internal OIDC connection details to this SPA. It relies on Quarkus OIDC to let SPA athenticate its users to OIDC and OAuth2 providers which may be technically challenging to support directly at the SPA level.
+
+Another user case for OIDC Proxy is to have several frontend https://quarkus.io/guides/security-oidc-code-flow-authentication[Quarkus OIDC web-app] endpoints to authenticate users using the same OIDC proxy configuration  before accessing the OIDC service endpoint.
```

Comment:
Frontends?

Reply:
@maxandersen I often refer to the Quarkus OIDC `web-app` endpoints as frontend (endpoints) because they interact with users directly via the browser (authenticate them by redirecting to OIDC providers, etc), but I removed this word to avoid any confusions